File: server/javabackend/src/main/java/com/gamevault/controller/AuthController.java

package com.gamevault.controller;

import com.gamevault.entity.Role;
import com.gamevault.entity.User;
import com.gamevault.entity.Setting;
import com.gamevault.entity.PasswordResetToken;
import com.gamevault.repository.UserRepository;
import com.gamevault.repository.PasswordResetTokenRepository;
import com.gamevault.repository.SettingRepository;
import com.gamevault.security.JwtTokenProvider;
import com.gamevault.service.AuditService;
import lombok.Data;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

@RestController
@RequestMapping("/api/auth")
public class AuthController {

    @Autowired
    AuthenticationManager authenticationManager;

    @Autowired
    UserRepository userRepository;
    
    @Autowired
    PasswordResetTokenRepository tokenRepository;
    
    @Autowired
    SettingRepository settingRepository;

    @Autowired
    PasswordEncoder passwordEncoder;

    @Autowired
    JwtTokenProvider tokenProvider;
    
    @Autowired
    AuditService auditService;

    @PostMapping("/signin")
    public ResponseEntity<?> authenticateUser(@RequestBody LoginRequest loginRequest) {
        Authentication authentication = authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        loginRequest.getUsername(),
                        loginRequest.getPassword()
                )
        );

        SecurityContextHolder.getContext().setAuthentication(authentication);
        String jwt = tokenProvider.generateToken(authentication);
        
        User user = userRepository.findByUsername(loginRequest.getUsername()).orElseThrow();
        
        auditService.log("LOGIN", "User logged in successfully", user.getUsername());

        Map<String, Object> response = new HashMap<>();
        response.put("token", jwt);
        response.put("user", user);

        return ResponseEntity.ok(response);
    }

    @PostMapping("/signup")
    public ResponseEntity<?> registerUser(@RequestBody SignUpRequest signUpRequest) {
        // Check configuration
        Setting regSetting = settingRepository.findByKey("ENABLE_REGISTRATION").orElse(null);
        if (regSetting != null && "false".equalsIgnoreCase(regSetting.getValue())) {
             return ResponseEntity.status(HttpStatus.FORBIDDEN).body("Registration is currently disabled by the administrator.");
        }

        if(userRepository.existsByUsername(signUpRequest.getUsername())) {
            return new ResponseEntity<>("Username is already taken!", HttpStatus.BAD_REQUEST);
        }

        if(userRepository.existsByEmail(signUpRequest.getEmail())) {
            return new ResponseEntity<>("Email Address already in use!", HttpStatus.BAD_REQUEST);
        }

        User user = new User();
        user.setUsername(signUpRequest.getUsername());
        user.setEmail(signUpRequest.getEmail());
        user.setPassword(passwordEncoder.encode(signUpRequest.getPassword()));
        user.setRole(Role.USER);

        userRepository.save(user);
        auditService.log("REGISTER", "New user registered: " + user.getUsername(), null);

        return new ResponseEntity<>("User registered successfully", HttpStatus.OK);
    }
    
    @PostMapping("/forgot-password")
    public ResponseEntity<?> forgotPassword(@RequestBody Map<String, String> request) {
        String email = request.get("email");
        if (email == null || email.isEmpty()) {
            return ResponseEntity.badRequest().body("Email is required");
        }

        // Logic: Try to find by email first, then by username
        userRepository.findByEmail(email)
                .or(() -> userRepository.findByUsername(email))
                .ifPresent(user -> {
                    String token = UUID.randomUUID().toString();
                    PasswordResetToken resetToken = new PasswordResetToken();
                    resetToken.setToken(token);
                    resetToken.setUser(user);
                    resetToken.setExpiryDate(LocalDateTime.now().plusHours(1));
                    tokenRepository.save(resetToken);
                    
                    // In production, send email. Here we log to system out for the admin to see.
                    System.out.println("=============================================");
                    System.out.println("PASSWORD RESET TOKEN FOR " + user.getEmail() + ": " + token);
                    System.out.println("=============================================");
                    
                    auditService.log("RESET_REQUEST", "Password reset requested for " + user.getEmail(), null);
                });
        
        // Always return OK to prevent email enumeration attacks
        return ResponseEntity.ok().build();
    }

    @PostMapping("/reset-password")
    public ResponseEntity<?> resetPassword(@RequestBody Map<String, String> request) {
        String token = request.get("token");
        String newPassword = request.get("newPassword");
        
        if (token == null || newPassword == null) {
             return ResponseEntity.badRequest().body("Token and new password are required");
        }

        return tokenRepository.findByToken(token)
                .filter(t -> !t.isExpired())
                .map(t -> {
                    User user = t.getUser();
                    user.setPassword(passwordEncoder.encode(newPassword));
                    userRepository.save(user);
                    tokenRepository.delete(t); // Invalidate token
                    auditService.log("RESET_SUCCESS", "Password reset successful for user " + user.getUsername(), null);
                    return ResponseEntity.ok().build();
                })
                .orElse(ResponseEntity.status(HttpStatus.BAD_REQUEST).body("Invalid or expired token"));
    }

    @Data
    static class LoginRequest {
        private String username;
        private String password;
    }

    @Data
    static class SignUpRequest {
        private String username;
        private String email;
        private String password;
    }
}

File: server/gobackend/controllers/auth_controller.go

package controllers

import (
	"gamevault/database"
	"gamevault/models"
	"net/http"
	"os"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

func SignIn(c *gin.Context) {
	var input struct {
		Username string `json:"username"`
		Password string `json:"password"`
	}
	if err := c.BindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var user models.User
	if err := database.DB.Where("username = ?", input.Username).First(&user).Error; err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid credentials"})
		return
	}

	if err := bcrypt.CompareHashAndPassword([]byte(user.Password), []byte(input.Password)); err != nil {
		c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid credentials"})
		return
	}

	if !user.Enabled {
		c.JSON(http.StatusForbidden, gin.H{"error": "Account disabled"})
		return
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS512, jwt.MapClaims{
		"sub":  user.Username,
		"role": user.Role,
		"exp":  time.Now().Add(time.Hour * 168).Unix(),
	})

	tokenString, _ := token.SignedString([]byte(os.Getenv("JWT_SECRET")))

	LogAudit("SYSTEM", "LOGIN", "User logged in", user.Username)

	c.JSON(http.StatusOK, gin.H{
		"token": tokenString,
		"user":  user,
	})
}

func SignUp(c *gin.Context) {
	// Check if registration is enabled
	var setting models.Setting
	if err := database.DB.Where("setting_key = ?", "ENABLE_REGISTRATION").First(&setting).Error; err == nil {
		if setting.Value == "false" {
			c.JSON(http.StatusForbidden, gin.H{"error": "Registration is currently disabled by the administrator."})
			return
		}
	}

	var input struct {
		Username string `json:"username"`
		Email    string `json:"email"`
		Password string `json:"password"`
	}
	if err := c.BindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(input.Password), bcrypt.DefaultCost)
	user := models.User{Username: input.Username, Email: input.Email, Password: string(hashedPassword)}

	if err := database.DB.Create(&user).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Username or Email already taken"})
		return
	}

	LogAudit("SYSTEM", "REGISTER", "New user registered", user.Username)
	c.JSON(http.StatusOK, gin.H{"message": "Registered successfully"})
}

func ForgotPassword(c *gin.Context) {
	var input struct {
		Email string `json:"email"`
	}
	if err := c.BindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var user models.User
	if err := database.DB.Where("email = ?", input.Email).First(&user).Error; err != nil {
		// Don't reveal user existence
		c.JSON(http.StatusOK, gin.H{"message": "If email exists, reset link sent"})
		return
	}

	token := uuid.New().String()
	resetToken := models.PasswordResetToken{
		Token:      token,
		UserID:     user.ID,
		ExpiryDate: time.Now().Add(1 * time.Hour),
	}
	database.DB.Create(&resetToken)

	// Mock Email Sending
	println("RESET TOKEN FOR " + user.Email + ": " + token)

	c.JSON(http.StatusOK, gin.H{"message": "Reset link sent"})
}

func ResetPassword(c *gin.Context) {
	var input struct {
		Token       string `json:"token"`
		NewPassword string `json:"newPassword"`
	}
	if err := c.BindJSON(&input); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}

	var resetToken models.PasswordResetToken
	if err := database.DB.Where("token = ?", input.Token).First(&resetToken).Error; err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid token"})
		return
	}

	if time.Now().After(resetToken.ExpiryDate) {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Token expired"})
		return
	}

	var user models.User
	database.DB.First(&user, resetToken.UserID)
	hashedPassword, _ := bcrypt.GenerateFromPassword([]byte(input.NewPassword), bcrypt.DefaultCost)
	user.Password = string(hashedPassword)
	database.DB.Save(&user)
	database.DB.Delete(&resetToken)

	c.JSON(http.StatusOK, gin.H{"message": "Password reset successful"})
}
